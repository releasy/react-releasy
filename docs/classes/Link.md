# Link

A class responsible to fetch data from server.

## Usage

```javascript
import { Link } from 'react-releasy';

const link = new Link({
  url: 'https://yourserveraddress.com/graphql',
  fetchTimeout: 30000,
  retryDelays: [1000, 3000, 5000, 10000],
  headers: {},
});
```

Also, we can use a function to generate the `headers` object:

```javascript
import { Link } from 'react-releasy';

// this function would get the token inside localStorage or AsyncStorage
import { getToken } from './auth';

const link = new Link({
  url: 'https://yourserveraddress.com/graphql',
  fetchTimeout: 30000,
  retryDelays: [1000, 3000, 5000, 10000],
  headers: () => ({ token: getToken() }),
});
```

## Props

- **url**: the server address
- **fetchTimeout**: how long to wait(in ms) for a success response *(optional)*
- **retryDelays**: how long to wait(in ms) before make a new request after a failure *(optional)*
- **headers**: additional headers that will be send to the server *(optional)*

----

[Go back](../)
