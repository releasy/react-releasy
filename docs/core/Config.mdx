---
name: Config
menu: Core
route: /core/config
---

# Config

A configuration class.

## Usage

```javascript
import { Config, InMemoryCache, Link } from 'react-releasy';

const config = new Config({
  link: new Link({ url: 'https://yourserveraddress.com/graphql' }),
  cache: new InMemoryCache(),
  devTools: true,
  networkLogger: true,
});
```
