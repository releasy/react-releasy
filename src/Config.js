// @flow
import invariant from 'invariant';

import type { CacheConfig, GraphQLResponse, RequestNode, UploadableMap, Variables } from 'relay-runtime';

import createEnvironment from './relay/createEnvironment';
import { setEnvironment } from './store';

export type LinkInterface = {|
  fetch: (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: UploadableMap,
  ) => ?Promise<any>,
|};

export type CacheInterface = {|
  get: (queryID: string, variables: Variables) => ?GraphQLResponse,
  set: (queryID: string, variables: Variables, payload: GraphQLResponse) => void,
  clear: () => void,
|};

type Props = {
  link: LinkInterface,
  cache?: CacheInterface,
  devTools?: boolean,
  networkLogger?: boolean,
};

class Config {
  link: LinkInterface;
  cache: ?CacheInterface;
  devTools: boolean;
  networkLogger: boolean;

  constructor(props: Props) {
    invariant(props, 'Config: missing props');
    invariant(props.link, 'Config: missing "link"');

    this.link = props.link;
    this.cache = props.cache || null;
    this.devTools = props.devTools || false;
    this.networkLogger = props.networkLogger || false;

    setEnvironment(createEnvironment(this));
  }
}

export default Config;
