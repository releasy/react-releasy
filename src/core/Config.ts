import { CacheConfig, RequestNode, UploadableMap, Variables, GraphQLResponse } from 'relay-runtime';
import invariant from 'invariant';

export interface LinkInterface {
  fetch: (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables: UploadableMap,
  ) => Promise<GraphQLResponse>;
}

export interface CacheInterface {
  get: (queryID: string, variables: Variables) => GraphQLResponse;
  set: (queryID: string, variables: Variables, payload: GraphQLResponse) => void;
  clear: () => void;
  size: () => number;
}

type ConfigPropsType = {
  link: LinkInterface,
  cache?: CacheInterface,
  devTools?: boolean,
  networkLogger?: boolean,
};

class Config {
  link: LinkInterface;
  cache: CacheInterface;
  devTools: boolean;
  networkLogger: boolean;

  constructor(props: ConfigPropsType) {
    invariant(props, 'Config: missing props');
    invariant(props.link, 'Config: missing "link"');

    this.link = props.link;
    this.cache = props.cache || null;
    this.devTools = props.devTools || false;
    this.networkLogger = props.networkLogger || false;
  }
}

export default Config;
