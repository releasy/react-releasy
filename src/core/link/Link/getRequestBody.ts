import { RequestNode, UploadableMap, Variables } from 'relay-runtime';

const getRequestBodyWithUploadables = (
  request: RequestNode,
  variables: Variables,
  uploadables: UploadableMap,
) => {
  const formData = new FormData();
  formData.append('query', request.text);
  formData.append('variables', JSON.stringify(variables));

  Object.keys(uploadables).forEach((key) => {
    formData.append(key, uploadables[key]);
  });

  return formData;
};

const getRequestBodyWithoutUplodables = (request: RequestNode, variables: Variables): string => {
  return JSON.stringify({
    variables,
    query: request.text,
  });
};

const getRequestBody = (
  request: RequestNode,
  variables: Variables,
  uploadables?: UploadableMap,
): FormData | string => {
  if (uploadables) {
    return getRequestBodyWithUploadables(request, variables, uploadables);
  }

  return getRequestBodyWithoutUplodables(request, variables);
};

export default getRequestBody;
