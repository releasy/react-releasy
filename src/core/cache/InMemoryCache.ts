import { Variables, GraphQLResponse } from 'relay-runtime';
import RelayQueryResponseCache from 'relay-runtime/lib/RelayQueryResponseCache.js';

import { CacheInterface } from '../Config';

type InMemoryCacheProps = {
  ttl?: number,
  size?: number,
};

class InMemoryCache implements CacheInterface {
  store: RelayQueryResponseCache;

  constructor(props?: InMemoryCacheProps) {
    this.store = new RelayQueryResponseCache({
      size: props && props.size ? props.size : 250,
      ttl: props && props.ttl ? props.ttl : 60 * 1000 * 5, // default = 5 minutes
    });
  }

  get = (queryID: string, variables: Variables): GraphQLResponse => {
    return this.store.get(queryID, variables);
  }

  set = (queryID: string, variables: Variables, payload: GraphQLResponse): void => {
    this.store.set(queryID, variables, payload);
  }

  clear = (): void => {
    this.store.clear();
  }
}

export default InMemoryCache;
