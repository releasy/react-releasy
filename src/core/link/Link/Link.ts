import { CacheConfig, RequestNode, UploadableMap, Variables, GraphQLResponse } from 'relay-runtime';
import invariant from 'invariant';
import fetchWithRetries from 'fbjs/lib/fetchWithRetries';

import { LinkInterface } from '../../Config';
import getHeaders from './getHeaders';
import getRequestBody from './getRequestBody';

export type LinkPropsType = {
  url: string,
  fetchTimeout?: number,
  retryDelays?: number[],
  headers?: Object | Function,
};

class Link implements LinkInterface {
  url: string;
  fetchTimeout: number;
  retryDelays: number[];
  headers: Object | Function;

  constructor(props: LinkPropsType) {
    invariant(props, 'Link: missing props');
    invariant(props.url, 'Link: missing "url"');

    this.url = props.url;
    this.headers = props.headers || {};
    this.fetchTimeout = props.fetchTimeout || 30000;
    this.retryDelays = props.retryDelays || [1000, 3000, 5000, 10000];
  }

  fetch = (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: UploadableMap,
  ): Promise<GraphQLResponse> => {
    return new Promise(async (resolve) => {
      const customHeaders = typeof this.headers === 'function' ? this.headers() : this.headers;

      const headers = {
        ...getHeaders(uploadables),
        ...customHeaders,
      };

      const body = getRequestBody(request, variables, uploadables);

      const response = await fetchWithRetries(this.url, {
        headers,
        body,
        method: 'POST',
        fetchTimeout: this.fetchTimeout,
        retryDelays: this.retryDelays,
      });

      const contentType = response.headers.get('content-type');
      if (contentType && contentType.indexOf('application/json') !== -1) {
        resolve(response.json());
        return;
      }

      resolve(response.text());
    });
  }
}

export default Link;
