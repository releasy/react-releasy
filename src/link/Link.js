// @flow
import invariant from 'invariant';
import fetchWithRetries from 'fbjs/lib/fetchWithRetries';

import type { CacheConfig, RequestNode, UploadableMap, Variables } from 'relay-runtime';

import getHeaders from './getHeaders';
import getRequestBody from './getRequestBody';

export type Props = {
  url: string,
  fetchTimeout?: number,
  retryDelays?: Array<number>,
  headers?: Object | () => Object,
};

class Link {
  url: string;
  fetchTimeout: number;
  retryDelays: Array<number>;
  headers: Object | () => Object;

  constructor(props: Props) {
    invariant(props, 'Link: missing props');
    invariant(props.url, 'Link: missing "url"');

    this.url = props.url;
    this.headers = props.headers || {};
    this.fetchTimeout = props.fetchTimeout || 30000;
    this.retryDelays = props.retryDelays || [1000, 3000, 5000, 10000];
  }

  fetch = async (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: UploadableMap,
  ): Promise<any> => {
    const customHeaders = typeof this.headers === 'function' ? this.headers() : this.headers;

    const headers = {
      ...getHeaders(uploadables),
      ...customHeaders,
    };

    const body = getRequestBody(request, variables, uploadables);

    const response = await fetchWithRetries(this.url, {
      method: 'POST',
      fetchTimeout: this.fetchTimeout,
      retryDelays: this.retryDelays,
      headers,
      body,
    });

    const contentType = response.headers.get('content-type');
    if (contentType && contentType.indexOf('application/json') !== -1) {
      return response.json();
    }

    return response.text();
  }
}

export default Link;
