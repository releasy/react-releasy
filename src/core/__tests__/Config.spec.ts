import InMemoryCache from '../cache/InMemoryCache';
import Link from '../link/Link';
import Config from '../Config';

it('should create a valid Config', () => {
  const link = new Link({
    url: '/graphql',
  });

  const config = new Config({
    link,
  });

  expect(config).toMatchSnapshot();
});

it('should create a Config with cache', () => {
  const link = new Link({
    url: '/graphql',
  });

  const cache = new InMemoryCache();

  const config = new Config({
    link,
    cache,
  });

  expect(config).toMatchSnapshot();
});
