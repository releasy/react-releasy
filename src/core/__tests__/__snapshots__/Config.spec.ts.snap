// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a Config with cache 1`] = `
Config {
  "cache": InMemoryCache {
    "clear": [Function],
    "get": [Function],
    "set": [Function],
    "size": [Function],
    "store": RelayQueryResponseCache {
      "_responses": Map {},
      "_size": 250,
      "_ttl": 300000,
    },
  },
  "devTools": false,
  "link": Link {
    "fetch": [Function],
    "fetchTimeout": 30000,
    "headers": Object {},
    "retryDelays": Array [
      1000,
      3000,
      5000,
      10000,
    ],
    "url": "/graphql",
  },
  "networkLogger": false,
  "ssrMode": false,
}
`;

exports[`should create a Config with ssrMode 1`] = `
Config {
  "cache": null,
  "devTools": false,
  "link": Link {
    "fetch": [Function],
    "fetchTimeout": 30000,
    "headers": Object {},
    "retryDelays": Array [
      1000,
      3000,
      5000,
      10000,
    ],
    "url": "/graphql",
  },
  "networkLogger": false,
  "ssrMode": true,
}
`;

exports[`should create a valid Config 1`] = `
Config {
  "cache": null,
  "devTools": false,
  "link": Link {
    "fetch": [Function],
    "fetchTimeout": 30000,
    "headers": Object {},
    "retryDelays": Array [
      1000,
      3000,
      5000,
      10000,
    ],
    "url": "/graphql",
  },
  "networkLogger": false,
  "ssrMode": false,
}
`;
